src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs

```diff
@@ -19,26 +22,136 @@ public abstract partial class BaselineTree
         private const ulong MaxNodeIndex = MaxNodes - 1;
         public const uint MaxLeafIndex = uint.MaxValue;
 
-        private readonly IKeyValueStore _keyValueStore;
+        private readonly IDb _db;
+        private readonly IKeyValueStore _metadataKeyValueStore;
         private readonly byte[] _dbPrefix;
+        private readonly object _synchroObject = new object();
 
         static BaselineTree()
         {
         }
 
-        public uint Count { get; set; }
+        public uint Count { get; private set; }
+
+        public long LastBlockWithLeaves { get; set; }
+
+        public Keccak LastBlockDbHash { get; set; } = Keccak.Zero;
 
         /* baseline does not use a sparse merkle tree - instead they use a single zero hash value
            does it expose any attack vectors? */
         internal static Keccak ZeroHash = Keccak.Zero;
 
-        public BaselineTree(IKeyValueStore keyValueStore, byte[] _dbPrefix, int truncationLength)
+        public BaselineTree(IDb db, IKeyValueStore metadataKeyValueStore, byte[] _dbPrefix, int truncationLength)
         {
             TruncationLength = truncationLength;
-            _keyValueStore = keyValueStore ?? throw new ArgumentNullException(nameof(keyValueStore));
+            _db = db ?? throw new ArgumentNullException(nameof(db));
+            _metadataKeyValueStore = metadataKeyValueStore ?? throw new ArgumentNullException(nameof(metadataKeyValueStore));
+
             this._dbPrefix = _dbPrefix;
+            InitializeMetadata();
+            Root = Count == 0 ? Keccak.Zero : LoadValue(new Index(0, 0));
+        }
+
+        private byte[] BuildDbKey(ulong nodeIndex)
+        {
+            return Rlp.Encode(Rlp.Encode(_dbPrefix), Rlp.Encode(nodeIndex)).Bytes;
+        }
+
+        private byte[] MetadataBuildDbKey(long blockNumber)
+        {
+            return Rlp.Encode(Rlp.Encode(_dbPrefix), Rlp.Encode(blockNumber)).Bytes;
+        }
+
+        private void SaveValue(in Index index, byte[] hashBytes)
+        {
+            _db[BuildDbKey(index.NodeIndex)] = hashBytes;
+        }
+
+        private Keccak LoadValue(in Index index)
+        {
+            byte[]? nodeHashBytes = _db[BuildDbKey(index.NodeIndex)];
+            if (nodeHashBytes == null)
+            {
+                return ZeroHash;
+            }
+
+            return new Keccak(nodeHashBytes);
+        }
+
+        public BatchWrite StartBatch()
+        {
+            return new BatchWrite(_synchroObject);
+        }
+
+        #region Metadata operation
+
+        private const long CurrentBlockIndex = -1;
```

Comment:
seems a bit strange to keep the block info in the BaselineTree class - why not a separate class?